File: glicko.py
import numpy as np

TAU = 0.5
MU = 1500.0
SCALE = 173.7178


def g(phi):
    return 1 / np.sqrt(1 + 3 * (phi**2) / (np.pi**2))


def E(mu, mu_j, phi_j):
    return 1 / (1 + np.exp(-g(phi_j) * (mu - mu_j)))


def compute_variance(mu, opponents):
    v_inv = 0
    for mu_j, phi_j, s_j in opponents:
        E_j = E(mu, mu_j, phi_j)
        v_inv += (g(phi_j) ** 2) * E_j * (1 - E_j)
    return 1 / v_inv


def compute_delta(mu, phi, v, opponents):
    delta_sum = 0
    for mu_j, phi_j, s_j in opponents:
        E_j = E(mu, mu_j, phi_j)
        delta_sum += g(phi_j) * (s_j - E_j)
    return v * delta_sum


def update_volatility(phi, delta, v, sigma):
    a = np.log(sigma**2)
    tau = TAU
    epsilon = 1e-6

    def f(x):
        exp_x = np.exp(x)
        num = exp_x * (delta**2 - phi**2 - v - exp_x)
        den = 2 * (phi**2 + v + exp_x) ** 2
        return num / den - (x - a) / (tau**2)

    A = a
    B = None
    if delta**2 > phi**2 + v:
        B = np.log(delta**2 - phi**2 - v)
    else:
        B = a - 1.0

    fA = f(A)
    fB = f(B)

    while abs(B - A) > epsilon:
        C = A + (A - B) * fA / (fB - fA)
        fC = f(C)

        if fC * fB < 0:
            A = B
            fA = fB
        else:
            fA /= 2

        B = C
        fB = fC

    return np.exp(A / 2)


def f(x, delta, phi, v, a):
    exp_x = np.exp(x)
    num = exp_x * (delta**2 - phi**2 - v - exp_x)
    den = 2 * (phi**2 + v + exp_x) ** 2
    return num / den - (x - a) / (TAU**2)


def glicko2_rating_update(player_rating, game_results):
    mu = (player_rating["rating"] - MU) / SCALE
    phi = player_rating["rd"] / SCALE
    sigma = player_rating["volatility"]

    if not game_results:
        phi_star = np.sqrt(phi**2 + sigma**2)
        return {"rating": MU + SCALE * mu, "rd": SCALE * phi_star, "volatility": sigma}

    opponents = []
    for result in game_results:
        opp_mu = (result["opponent_rating"] - MU) / SCALE
        opp_phi = result["opponent_rd"] / SCALE
        s = result["score"]
        opponents.append((opp_mu, opp_phi, s))

    v = compute_variance(mu, opponents)
    delta = compute_delta(mu, phi, v, opponents)
    sigma_prime = update_volatility(phi, delta, v, sigma)
    phi_star = np.sqrt(phi**2 + sigma_prime**2)
    phi_prime = 1 / np.sqrt(1 / (phi_star**2) + 1 / v)

    sum_term = 0
    for mu_j, phi_j, s_j in opponents:
        E_j = E(mu, mu_j, phi_j)
        sum_term += g(phi_j) * (s_j - E_j)

    mu_prime = mu + phi_prime**2 * sum_term

    return {
        "rating": MU + SCALE * mu_prime,
        "rd": SCALE * phi_prime,
        "volatility": sigma_prime,
    }


if __name__ == "__main__":
    # test should result in {'rating': 1464.06, 'rd': 151.52, 'volatility': 0.05999}
    player_rating = {"rating": 1500, "rd": 200, "volatility": 0.06}

    game_results = [
        {"opponent_rating": 1400, "opponent_rd": 30, "score": 1},
        {"opponent_rating": 1550, "opponent_rd": 100, "score": 0},
        {"opponent_rating": 1700, "opponent_rd": 300, "score": 0},
    ]

    updated_rating = glicko2_rating_update(player_rating, game_results)
    print(updated_rating)

File: src/glicko.rs
const TAU: f64 = 1.0;
const MU: f64 = 1500.0;
const SCALE: f64 = 173.7178;

#[derive(Debug, Clone, serde::Serialize)]
pub struct Player {
    pub rating: f64,
    pub rd: f64,
    pub volatility: f64,
    pub matches: i32,
}

#[derive(Debug, Clone)]
pub struct Opponent {
    pub rating: f64,
    pub rd: f64,
    pub score: f64,
}

fn g(phi: f64) -> f64 {
    1.0 / (1.0 + 3.0 * phi.powi(2) / std::f64::consts::PI.powi(2)).sqrt()
}

fn e(mu: f64, mu_j: f64, phi_j: f64) -> f64 {
    1.0 / (1.0 + (-g(phi_j) * (mu - mu_j)).exp())
}

fn compute_variance(mu: f64, opponents: &[(f64, f64, f64)]) -> f64 {
    let mut v_inv = 0.0;
    for &(mu_j, phi_j, _s_j) in opponents {
        let e_j = e(mu, mu_j, phi_j);
        v_inv += g(phi_j).powi(2) * e_j * (1.0 - e_j);
    }
    1.0 / v_inv
}

fn compute_delta(mu: f64, v: f64, opponents: &[(f64, f64, f64)]) -> f64 {
    let mut delta_sum = 0.0;
    for &(mu_j, phi_j, s_j) in opponents {
        let e_j = e(mu, mu_j, phi_j);
        delta_sum += g(phi_j) * (s_j - e_j);
    }
    v * delta_sum
}

fn f_function(x: f64, delta: f64, phi: f64, v: f64, a: f64) -> f64 {
    let exp_x = x.exp();
    let num = exp_x * (delta.powi(2) - phi.powi(2) - v - exp_x);
    let den = 2.0 * (phi.powi(2) + v + exp_x).powi(2);
    num / den - (x - a) / TAU.powi(2)
}

fn update_volatility(phi: f64, delta: f64, v: f64, sigma: f64) -> f64 {
    let a = (sigma.powi(2)).ln();
    let epsilon = 1e-6;

    let mut a_val = a;
    let mut b_val = if delta.powi(2) > phi.powi(2) + v {
        (delta.powi(2) - phi.powi(2) - v).ln()
    } else {
        a - 1.0
    };

    let mut fa = f_function(a_val, delta, phi, v, a);
    let mut fb = f_function(b_val, delta, phi, v, a);

    while (b_val - a_val).abs() > epsilon {
        let c = a_val + (a_val - b_val) * fa / (fb - fa);
        let fc = f_function(c, delta, phi, v, a);

        if fc * fb < 0.0 {
            a_val = b_val;
            fa = fb;
        } else {
            fa /= 2.0;
        }

        b_val = c;
        fb = fc;
    }

    (a_val / 2.0).exp()
}

impl Player {
    pub fn update(&self, game_results: &[Opponent]) -> Player {
        let mu = (self.rating - MU) / SCALE;
        let phi = self.rd / SCALE;
        let sigma = self.volatility;

        if game_results.is_empty() {
            let phi_star = (phi.powi(2) + sigma.powi(2)).sqrt();
            return Player {
                rating: MU + SCALE * mu,
                rd: SCALE * phi_star,
                volatility: sigma,
                matches: self.matches,
            };
        }

        let opponents: Vec<(f64, f64, f64)> = game_results
            .iter()
            .map(|result| {
                let opp_mu = (result.rating - MU) / SCALE;
                let opp_phi = result.rd / SCALE;
                let s = result.score;
                (opp_mu, opp_phi, s)
            })
            .collect();

        let v = compute_variance(mu, &opponents);
        let delta = compute_delta(mu, v, &opponents);
        let sigma_prime = update_volatility(phi, delta, v, sigma);
        let phi_star = (phi.powi(2) + sigma_prime.powi(2)).sqrt();
        let phi_prime = 1.0 / (1.0 / phi_star.powi(2) + 1.0 / v).sqrt();

        let mut sum_term = 0.0;
        for &(mu_j, phi_j, s_j) in &opponents {
            let e_j = e(mu, mu_j, phi_j);
            sum_term += g(phi_j) * (s_j - e_j);
        }

        let mu_prime = mu + phi_prime.powi(2) * sum_term;

        Player {
            rating: MU + SCALE * mu_prime,
            rd: SCALE * phi_prime,
            volatility: sigma_prime,
            matches: self.matches + game_results.len() as i32,
        }
    }

    pub fn update_mut(&mut self, game_results: &[Opponent]) {
        *self = self.update(game_results);
    }
}

pub fn win_probability(rating_a: f64, rating_b: f64, rd_b: f64) -> f64 {
    let mu_a = (rating_a - MU) / SCALE;
    let mu_b = (rating_b - MU) / SCALE;
    let phi_b = rd_b / SCALE;

    let g_phi = 1.0 / (1.0 + 3.0 * phi_b.powi(2) / std::f64::consts::PI.powi(2)).sqrt();
    let exponent = -g_phi * (mu_a - mu_b);
    let e_a = 1.0 / (1.0 + exponent.exp());

    e_a
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rating_update() {
        let player_rating = Player {
            rating: 1500.0,
            rd: 200.0,
            volatility: 0.06,
            matches: 0,
        };

        let game_results = vec![
            Opponent {
                rating: 1400.0,
                rd: 30.0,
                score: 1.0,
            },
            Opponent {
                rating: 1550.0,
                rd: 100.0,
                score: 0.0,
            },
            Opponent {
                rating: 1700.0,
                rd: 300.0,
                score: 0.0,
            },
        ];

        let updated_rating = player_rating.update(&game_results);

        // Expected result: {'rating': 1464.06, 'rd': 151.52, 'volatility': 0.05999}
        println!("Updated rating: {:?}", updated_rating);
        assert!((updated_rating.rating - 1464.06).abs() < 0.1);
        assert!((updated_rating.rd - 151.52).abs() < 0.1);
        assert!((updated_rating.volatility - 0.05999).abs() < 0.001);
    }

    #[test]
    fn test_win_probability() {
        let prob = win_probability(1122.0, 1976.0, 111.0);
        println!("Win probability: {}", prob);
    }
}
